### PR TITLE
Fix checks for serving AMP response and improve AMP compatibility

### LINKED ISSUE #9588: Check for /amp/ suffix when determining is_amp_request
As per @westonruter's [comment](https://github.com/Automattic/jetpack/pull/9458#pullrequestreview-121498378) on the AMP support patch.

> Until Automattic/amp-wp#1148 is implemented, this should also check if the URL path ends in /amp/. Otherwise, AMP URLs on sites with amp theme support like https://make.xwp.co/2018/05/03/wordpress-amp-plugin-0-7-release/amp/ won't get recognized as an AMP request here.

### LINKED ISSUE #10973: Missing Social Icons using AMP and WP 5.0.x
**Describe the bug**
There seems to be an issue with the recent changes in WP 5.0.1 with AMP and Jetpack's social sharing icons that causes them to no longer display.

**To Reproduce**
1. Scroll down to the bottom of the post where the social sharing links are
2. See error

**Expected behavior**
Social Sharing icons should be displayed instead of basic ul li list

**Screenshots**
screenshot

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser Chrome
 - Version 70.0.3538.110

### PR DESCRIPTION
**Pull request by gravityrail**
Fixes #9588
Fixes #9588
Fixes #10973

Based on #10918

#### Changes proposed in this Pull Request:
 
* Make AMP checks later in boot sequence - AMP 1.0 requires that `is_amp_endpoint()` be called after `parse_query`.

This avoids messages like this being printed to the screen or debug.log:

```
[11-Dec-2018 18:10:52 UTC] PHP Notice:  is_amp_endpoint was called <strong>incorrectly</strong>. is_amp_endpoint() was called before the &#039;parse_query&#039; hook was called. This function will always return &#039;false&#039; before the &#039;parse_query&#039; hook is called. Please see <a href="https://codex.wordpress.org/Debugging_in_WordPress">Debugging in WordPress</a> for more information. (This message was added in version 0.4.2.) in /srv/www/wordpress-default/public_html/wp-includes/functions.php on line 4169
```

It also ensures that the stats pixel is loaded

#### Testing instructions:
* Ensure `WP_DEBUG` and `WP_DEBUG_DISPLAY` are enabled
* Ensure AMP 1.x is installed and active
* Configure AMP for "Native" mode (should also try with "Paired" mode and "Classic" mode)
* Enable Jetpack features and load pages with those features
* Check that no errors are printed, and features are either working or hidden
* Load pages with AMP disabled (or load regular pages with AMP not in "Native" mode) and ensure features still work correctly

Good features to check: Carousel, Likes, Gallery, social widgets

#### Proposed changelog entry for your changes:

* Compatibility improvements with AMP plugin 1.0 (working stats pixel, remove warnings)

cc @westonruter

### CODE REVIEW

**Review by westonruter — DISMISSED** [review]
I deployed the changes to https://weston.ruter.net/ and I didn't notice any regressions or PHP log entries, though I don't have all modules active on my site.

**Comment by westonruter** [comment]
Note that this PR does not fix #9730 and #10314.

**Comment by gravityrail** [comment]
Whoops! Thanks @westonruter

**Comment by westonruter** [comment]
Note that images on AMP pages are currently broken with the current version of Jetpack because lazy images aren't being disabled. This PR should be marked important, I believe.

**Comment by zinigor** [comment]
I have tested this one a bit, seems to be working fine, but I have a question. Can we just filter out modules instead of introducing many checks into their code? We can have something like this for a predefined list of modules that shouldn't even be loaded if it's an AMP render:

```
	static function filter_modules( $modules ) {
		if ( ! self::is_amp_request() ) {
			return $modules;
		}

		l( array_diff( $modules, array( 'lazy-images' ) ) );
		return array_diff( $modules, array( 'lazy-images' ) );
	}
```

We can have this as a static method in `class.jetpack-amp-support.php` and hook it onto the [`jetpack_active_modules`](https://developer.jetpack.com/hooks/jetpack_active_modules/) filter.

**Comment by westonruter** [comment]
> I have tested this one a bit, seems to be working fine, but I have a question. Can we just filter out modules instead of introducing many checks into their code?

@zinigor I do not believe so because `is_amp_endpoint()` only works after the `parse_query` action. I assume that the filtering of the modules happens _much_ earlier than that action, correct?

**Comment by zinigor** [comment]
Oh, yes, you're right, it runs on `plugins_loaded`. OK then, multiple conditions work too.

**Comment by zinigor** [comment]
I have found that with Masterbar enabled a logged in view seems to be way off, so we need to disable Masterbar as well. For that I had to change the way the module is initialized, @vindl  can you please take a look and tell us if there are any caveats to doing this? The admin bar seems to work OK in non-AMP requests, but still, I'd love a second set of eyes.

**Comment by zinigor** [comment]
Meanwhile I'm going to work on the `wpcom` build failure.

**Comment by gravityrail** [comment]
@zinigor Just to confirm, yes, the reason I filter functionality rather than modules is because when Jetpack is loading its modules, we don't know yet whether the page is going to be rendered as AMP.

**Review by jeherve — DISMISSED** [review]
It seems to work well for me. I just added back something for the Likes, I believe it had been removed by mistake.

**Comment by dereksmart** [comment]
I'm not sure what I'm doing wrong to test this. When I am running the AMP v1.01, I get a gang of notices/warnings immediately on activation, and any time I try to change the settings there. 

image

With AMP in native mode, many of the Jetpack features do not work properly. 
- Likes: completely missing from the post
- Sharing buttons: completely different styling 
- Extra Sidebar Widgets (example: social icons): CSS is stripped completely. 

@jeherve @zinigor how did you both test this successfully?

**Comment by westonruter** [comment]
@dereksmart What version of PHP are you running? @kraftbj also ran into an issue with `INTL_*` constants in https://github.com/ampproject/amp-wp/issues/1439 and https://github.com/ampproject/amp-wp/pull/1440 so that may be a clue. The version of ICU may be older than is expected. The logic here may need to be improved, by checking for the ICU version somehow:

https://github.com/ampproject/amp-wp/blob/0b085ae2a6bf8fed6f5482589aa2aeca25ec0c68/includes/class-amp-http.php#L203-L209

> Likes: completely missing from the post

It is expected that Likes are missing because they are not supported in AMP yet. This is a known issue. See #9730 and #9555.

> Sharing buttons: completely different styling

This is also expected, since the sharing buttons use `amp-social-share` components. Perhaps styling can be better aligned, but it's not a regression introduced by this PR.

> Extra Sidebar Widgets (example: social icons): CSS is stripped completely.

This could be that AMP is stripping out the extra CSS due to going over the 50KB limit. If you look for the `<style amp-custom>` element in the page there is an HTML comment that precedes it with a manifest of the stylesheets that comprise the AMP styles. Do you see it being excluded there?

**Comment by dereksmart** [comment]
@westonruter 

PHP Version: `PHP 7.2.13`, which indeed [does look](http://php.net/manual/en/migration72.deprecated.php#migration72.deprecated.INTL_IDNA_VARIANT_2003-variant) to be deprecated in.

> Likes & Sharing 
- Thank you for clarifying this is not a regression in this PR. The testing instruction asked to test `Good features to check: Carousel, Likes, Gallery, social widgets`. But I assume I am only meant to check if `PHP Notice:  is_amp_endpoint` is there? 

> This could be that AMP is stripping out the extra CSS due to going over the 50KB limit.
Confirmed this is the case. Thank you for showing me how to check. Is this something known as well?

**Comment by westonruter** [comment]
> PHP Version: PHP 7.2.13, which indeed does look to be deprecated in.

I'm not sure what to do there if ICU itself is outdated.

> But I assume I am only meant to check if PHP Notice: is_amp_endpoint is there?

Correct. If you can confirm that there is no PHP notice.

> Confirmed this is the case. Thank you for showing me how to check. Is this something known as well?

In the AMP settings page, can you try turning off the admin bar?

image

**Comment by westonruter** [comment]
@dereksmart and you're looking at a `singular` post?

**Comment by dereksmart** [comment]
Ok, so testing with `classic` I was able to reproduce the CSS problem with the sharing icons, and verified that this PR fixes it. 

However, I was unable to find any notices in master, but I do see a notice with this patch applied. 

https://dereksmart.wpsandbox.me/guten-gallery/amp/

_send for help_

**Comment by westonruter** [comment]
> However, I was unable to find any notices in master, but I do see a notice with this patch applied.

@dereksmart This should be easily fixed by applying the following patch to `master/modules/comment-likes.php`:

```diff
- add_action( 'init', array( $this, 'frontend_init' ) );
+ add_action( 'template_redirect', array( $this, 'frontend_init' ) );
```

The `frontend_init` method has no need to run so early. Also, by switching from `init` to `template_redirect` then there is no need for the `is_admin()` conditional in it.

**Comment by dereksmart** [comment]
@westonruter nice catch, pushed that to this branch. 

Activated all other modules and found no notices while loading a singular page in classic mode. 

Is it expected that the gallery does not actually display as a gallery in AMP view? https://dereksmart.wpsandbox.me/guten-gallery/amp/

**Comment by dereksmart** [comment]
gallery (my comment above) is not a regression anyway. I think this is good to ship.

**Review by dereksmart — APPROVED** [review]

**Review by jeherve — APPROVED** [review]
Works well for me too. Merging.

**Comment by westonruter** [comment]
@dereksmart And no notices are generated in native/paired mode either, correct?

**Comment by dereksmart** [comment]
@westonruter actually I do see some with Native enabled. IDK how I missed it. 

> ( ! ) Notice: Trying to get property 'post_type' of non-object in /home/dereksma/public_html/wp-content/plugins/amp/includes/class-amp-post-type-support.php on line 80

> ( ! ) Notice: Trying to get property 'ID' of non-object in /home/dereksma/public_html/wp-content/plugins/amp/includes/class-amp-post-type-support.php on line 101

**Comment by dereksmart** [comment]
I've confirmed these new notices are happening when Jetpack's wpcom toolbar feature is active.

**Comment by westonruter** [comment]
@dereksmart do you have stack traces for those?

**Comment by dereksmart** [comment]
@westonruter 

[03-Jan-2019 23:02:26 UTC] PHP Notice:  Trying to get property 'post_type' of non-object in /home/dereksma/public_html/wp-content/plugins/amp/includes/class-amp-post-type-support.php on line 80
[03-Jan-2019 23:02:26 UTC] PHP Stack trace:
[03-Jan-2019 23:02:26 UTC] PHP   1. {main}() /home/dereksma/public_html/index.php:0
[03-Jan-2019 23:02:26 UTC] PHP   2. require() /home/dereksma/public_html/index.php:17
[03-Jan-2019 23:02:26 UTC] PHP   3. wp() /home/dereksma/public_html/wp-blog-header.php:16
[03-Jan-2019 23:02:26 UTC] PHP   4. WP->main() /home/dereksma/public_html/wp-includes/functions.php:1091
[03-Jan-2019 23:02:26 UTC] PHP   5. WP->query_posts() /home/dereksma/public_html/wp-includes/class-wp.php:739
[03-Jan-2019 23:02:26 UTC] PHP   6. WP_Query->query() /home/dereksma/public_html/wp-includes/class-wp.php:622
[03-Jan-2019 23:02:26 UTC] PHP   7. WP_Query->get_posts() /home/dereksma/public_html/wp-includes/class-wp-query.php:3365
[03-Jan-2019 23:02:26 UTC] PHP   8. WP_Query->parse_query() /home/dereksma/public_html/wp-includes/class-wp-query.php:1701
[03-Jan-2019 23:02:26 UTC] PHP   9. do_action_ref_array() /home/dereksma/public_html/wp-includes/class-wp-query.php:1050
[03-Jan-2019 23:02:26 UTC] PHP  10. WP_Hook->do_action() /home/dereksma/public_html/wp-includes/plugin.php:531
[03-Jan-2019 23:02:26 UTC] PHP  11. WP_Hook->apply_filters() /home/dereksma/public_html/wp-includes/class-wp-hook.php:310
[03-Jan-2019 23:02:26 UTC] PHP  12. jetpack_initialize_masterbar() /home/dereksma/public_html/wp-includes/class-wp-hook.php:286
[03-Jan-2019 23:02:26 UTC] PHP  13. Jetpack_AMP_Support::is_amp_request() /home/dereksma/public_html/wp-content/plugins/jetpack/modules/masterbar.php:23
[03-Jan-2019 23:02:26 UTC] PHP  14. is_amp_endpoint() /home/dereksma/public_html/wp-content/plugins/jetpack/3rd-party/class.jetpack-amp-support.php:55
[03-Jan-2019 23:02:26 UTC] PHP  15. AMP_Theme_Support::get_template_availability() /home/dereksma/public_html/wp-content/plugins/amp/includes/amp-helper-functions.php:286
[03-Jan-2019 23:02:26 UTC] PHP  16. AMP_Post_Type_Support::get_support_errors() /home/dereksma/public_html/wp-content/plugins/amp/includes/class-amp-theme-support.php:589
[03-Jan-2019 23:02:26 UTC] PHP Notice:  Trying to get property 'ID' of non-object in /home/dereksma/public_html/wp-content/plugins/amp/includes/class-amp-post-type-support.php on line 101
[03-Jan-2019 23:02:26 UTC] PHP Stack trace:
[03-Jan-2019 23:02:26 UTC] PHP   1. {main}() /home/dereksma/public_html/index.php:0
[03-Jan-2019 23:02:26 UTC] PHP   2. require() /home/dereksma/public_html/index.php:17
[03-Jan-2019 23:02:26 UTC] PHP   3. wp() /home/dereksma/public_html/wp-blog-header.php:16
[03-Jan-2019 23:02:26 UTC] PHP   4. WP->main() /home/dereksma/public_html/wp-includes/functions.php:1091
[03-Jan-2019 23:02:26 UTC] PHP   5. WP->query_posts() /home/dereksma/public_html/wp-includes/class-wp.php:739
[03-Jan-2019 23:02:26 UTC] PHP   6. WP_Query->query() /home/dereksma/public_html/wp-includes/class-wp.php:622
[03-Jan-2019 23:02:26 UTC] PHP   7. WP_Query->get_posts() /home/dereksma/public_html/wp-includes/class-wp-query.php:3365
[03-Jan-2019 23:02:26 UTC] PHP   8. WP_Query->parse_query() /home/dereksma/public_html/wp-includes/class-wp-query.php:1701
[03-Jan-2019 23:02:26 UTC] PHP   9. do_action_ref_array() /home/dereksma/public_html/wp-includes/class-wp-query.php:1050
[03-Jan-2019 23:02:26 UTC] PHP  10. WP_Hook->do_action() /home/dereksma/public_html/wp-includes/plugin.php:531
[03-Jan-2019 23:02:26 UTC] PHP  11. WP_Hook->apply_filters() /home/dereksma/public_html/wp-includes/class-wp-hook.php:310
[03-Jan-2019 23:02:26 UTC] PHP  12. jetpack_initialize_masterbar() /home/dereksma/public_html/wp-includes/class-wp-hook.php:286
[03-Jan-2019 23:02:26 UTC] PHP  13. Jetpack_AMP_Support::is_amp_request() /home/dereksma/public_html/wp-content/plugins/jetpack/modules/masterbar.php:23
[03-Jan-2019 23:02:26 UTC] PHP  14. is_amp_endpoint() /home/dereksma/public_html/wp-content/plugins/jetpack/3rd-party/class.jetpack-amp-support.php:55
[03-Jan-2019 23:02:26 UTC] PHP  15. AMP_Theme_Support::get_template_availability() /home/dereksma/public_html/wp-content/plugins/amp/includes/amp-helper-functions.php:286
[03-Jan-2019 23:02:26 UTC] PHP  16. AMP_Post_Type_Support::get_support_errors() /home/dereksma/public_html/wp-content/plugins/amp/includes/class-amp-theme-support.php:589

**Comment by westonruter** [comment]
@dereksmart Thanks. It seems Jetpack is doing the right thing:

https://github.com/Automattic/jetpack/blob/c7495918eea18c9fc44788f6bbcc2f0f7a62dbf8/modules/masterbar.php#L16-L26

Here's the relevant code in the AMP plugin:

https://github.com/ampproject/amp-wp/blob/96b194da2f63234f37748e753f0434614983a2dc/includes/class-amp-theme-support.php#L581-L589

It seems that `$query->is_singular() || $query->is_posts_page` is true, and yet `$query->get_queried_object()` is returning `null`. So then this code in the AMP plugin is raising notices:

https://github.com/ampproject/amp-wp/blob/96b194da2f63234f37748e753f0434614983a2dc/includes/class-amp-post-type-support.php#L74-L101

I'm confused why `is_singular` would be true but the `queried_object` seems to be `null`.

**Comment by westonruter** [comment]
Seems the problem is that `parse_query` happens before the query actually has been made, so `$wp_query->post` is empty. If this is the case, then really `is_amp_endpoint` should be disallowed before the `wp` action not the `parse_query` action.

@dereksmart If you do:

```diff
- add_action( 'parse_query', 'jetpack_initialize_masterbar', 99 ); 
+ add_action( 'wp', 'jetpack_initialize_masterbar', 99 ); 
```

Does that have the intended result in paired mode, where the AMP version continues to hide the masterbar but in the non-AMP version, the masterbar is shown?

**Comment by westonruter** [comment]
It looks like we'll have to make these changes to the AMP plugin as well, to warn when calling before `wp` rather than before `parse_query`:

```diff
diff --git a/includes/amp-helper-functions.php b/wp-content/plugins/amp/includes/amp-helper-functions.php
index 938eb40b..4a29b021 100644
--- a/includes/amp-helper-functions.php
+++ b/includes/amp-helper-functions.php
@@ -252,10 +252,10 @@ function is_amp_endpoint() {
 		return false;
 	}
 
-	$did_parse_query = did_action( 'parse_query' );
+	$did_parse_query = did_action( 'wp' );
 
 	if ( ! $did_parse_query ) {
-		_doing_it_wrong( __FUNCTION__, sprintf( esc_html__( "is_amp_endpoint() was called before the 'parse_query' hook was called. This function will always return 'false' before the 'parse_query' hook is called.", 'amp' ) ), '0.4.2' );
+		_doing_it_wrong( __FUNCTION__, sprintf( esc_html__( "is_amp_endpoint() was called before the 'wp' hook was called. This function will always return 'false' before the 'wp' hook is called.", 'amp' ) ), '0.4.2' );
 	}
 
 	$has_amp_query_var = (
diff --git a/includes/class-amp-post-type-support.php b/wp-content/plugins/amp/includes/class-amp-post-type-support.php
index 3863b1f2..a1dad0c4 100644
--- a/includes/class-amp-post-type-support.php
+++ b/includes/class-amp-post-type-support.php
@@ -75,6 +75,9 @@ class AMP_Post_Type_Support {
 		if ( ! ( $post instanceof WP_Post ) ) {
 			$post = get_post( $post );
 		}
+		if ( ! $post ) {
+			return array();
+		}
 		$errors = array();
 
 		if ( ! post_type_supports( $post->post_type, self::SLUG ) ) {
diff --git a/includes/class-amp-theme-support.php b/wp-content/plugins/amp/includes/class-amp-theme-support.php
index 215ce42c..97357edb 100644
--- a/includes/class-amp-theme-support.php
+++ b/includes/class-amp-theme-support.php
@@ -586,10 +586,12 @@ class AMP_Theme_Support {
 			 * @var WP_Post $queried_object
 			 */
 			$queried_object = $query->get_queried_object();
-			$support_errors = AMP_Post_Type_Support::get_support_errors( $queried_object );
-			if ( ! empty( $support_errors ) ) {
-				$matching_template['errors']    = array_merge( $matching_template['errors'], $support_errors );
-				$matching_template['supported'] = false;
+			if ( $queried_object ) {
+				$support_errors = AMP_Post_Type_Support::get_support_errors( $queried_object );
+				if ( ! empty( $support_errors ) ) {
+					$matching_template['errors']    = array_merge( $matching_template['errors'], $support_errors );
+					$matching_template['supported'] = false;
+				}
 			}
 		}
```

**Comment by jeherve** [comment]
I moved this remaining problem to a new issue: https://github.com/Automattic/jetpack/issues/11082

**Comment by dereksmart** [comment]
Thanks for digging @westonruter! 

`add_action( 'wp', 'jetpack_initialize_masterbar', 99 );`

indeed seems to work. Masterbar is there, and with no notices.

File: modules/stats.php [hunk 1]
```diff
@@ -983,7 +983,15 @@ function stats_admin_bar_menu( &$wp_admin_bar ) {
 
 	$title = esc_attr( __( 'Views over 48 hours. Click for more Site Stats.', 'jetpack' ) );
 
-	$menu = array( 'id' => 'stats', 'title' => "<div><script type='text/javascript'>var src;if(typeof(window.devicePixelRatio)=='undefined'||window.devicePixelRatio<2){src='$img_src';}else{src='$img_src_2x';}document.write('<img src=\''+src+'\' alt=\'$alt\' title=\'$title\' />');</script></div>", 'href' => $url );
+	$menu = array(
+		'id'   => 'stats',
+		'href' => $url,
+	);
+	if ( Jetpack_AMP_Support::is_amp_request() ) {
+		$menu['title'] = "<amp-img src='$img_src_2x' width=112 height=24 layout=fixed alt='$alt' title='$title'></amp-img>";
```
Comment (westonruter) [inline modules/stats.php]: This is made obsolete by #13450.

**Pull request merged**